Code/Geometry/Trajectory.cpp

```diff
@@ -0,0 +1,195 @@
+//
+// Copyright (C) 2003-2016 Paolo Tosco
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <RDGeneral/Invariant.h>
+#include <RDGeneral/BadFileException.h>
+#include <fstream>
+#include <sstream>
+#include <set>
+#include "Trajectory.h"
+
+namespace RDGeom {
+
+Snapshot::Snapshot(boost::shared_array<double> pos, double energy) :
+  d_trajectory(NULL),
+  d_energy(energy),
+  d_pos(pos) {}
+
+Point2D Snapshot::getPoint2D(unsigned int pointNum) const {
+  PRECONDITION(d_pos, "d_pos must not be NULL");
+  PRECONDITION(d_trajectory, "d_trajectory must not be NULL");
+  PRECONDITION(d_trajectory->dimension() == 2, "d_dimension must be == 2");
+  PRECONDITION(d_trajectory->numPoints(), "d_numPoints must be > 0");
+  URANGE_CHECK(pointNum, d_trajectory->numPoints() - 1);
+  unsigned int i = pointNum * d_trajectory->dimension();
+  return Point2D(d_pos[i], d_pos[i + 1]);
+}
+
+Point3D Snapshot::getPoint3D(unsigned int pointNum) const {
+  PRECONDITION(d_pos, "d_pos must not be NULL");
+  PRECONDITION(d_trajectory, "d_trajectory must not be NULL");
+  PRECONDITION(d_trajectory->dimension() >= 2, "d_dimension must be >= 2");
+  PRECONDITION(d_trajectory->numPoints(), "d_numPoints must be > 0");
+  URANGE_CHECK(pointNum, d_trajectory->numPoints() - 1);
+  unsigned int i = pointNum * d_trajectory->dimension();
+  return (Point3D(d_pos[i], d_pos[i + 1],
+          (d_trajectory->dimension() == 3) ? d_pos[i + 2] : 0.0));
+}
+
+Trajectory::Trajectory(unsigned int dimension, unsigned int numPoints) :
+  d_dimension(dimension),
+  d_numPoints(numPoints) {}
+
+Trajectory::Trajectory(const Trajectory &other) :
+  d_dimension(other.d_dimension),
+  d_numPoints(other.d_numPoints) {
+  for (SnapshotSPtrVect::const_iterator vectIt = other.d_snapshotVect.begin();
+    vectIt != other.d_snapshotVect.end(); ++vectIt)
+    addSnapshot(new Snapshot(*(*vectIt)));
+}
+
+unsigned int Trajectory::addSnapshot(Snapshot *s) {
+  return insertSnapshot(d_snapshotVect.size(), s);
+}
+
+Snapshot *Trajectory::getSnapshot(unsigned int snapshotNum) const {
+  URANGE_CHECK(snapshotNum, d_snapshotVect.size() - 1);
+  return d_snapshotVect[snapshotNum].get();
+}
+
+unsigned int Trajectory::insertSnapshot(unsigned int snapshotNum, Snapshot *s) {
+  URANGE_CHECK(snapshotNum, d_snapshotVect.size());
+  s->d_trajectory = this;
```

Comment:
A more practical question is now what happens when the destructors of T1 and T2 are called (since T1 may have already freed S1). You'd need to be a lot more careful.